rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -483,38 +490,36 @@ private module CallExprBaseMatchingInput implements MatchingInputSig {
     }
   }
 
-  pragma[nomagic]
-  private Type inferAnnotatedTypeInclSelf(AstNode n, TypePath path) {
-    result = getTypeAnnotation(n).resolveTypeAtInclSelf(path)
-  }
-
   private class FunctionDecl extends Declaration, Function {
-    override TypeParam getATypeParam() { result = this.getGenericParamList().getATypeParam() }
+    override TypeParameter getTypeParameter(TypeParameterPosition ppos) {
+      typeParamMatchPosition(this.getGenericParamList().getATypeParam(), result, ppos)
+      or
+      exists(TraitItemNode trait | this = trait.getAnAssocItem() |
+        typeParamMatchPosition(trait.getTypeParam(_), result, ppos)
+        or
+        ppos.isSelf() and result = TSelfTypeParameter(trait)
```

Comment:
So, the `Self` type parameter is now moved from the trait into each of the methods, which makes sense. But I wonder if this approach will also work once [support for associated type defaults](https://github.com/rust-lang/rust/issues/29661) lands? E.g., in something like
```rust
trait MyTrait {
  type T = Option<Self>

  fn foo(self) -> T;
}
```

Perhaps it will Just Work with your other PR that introduced support for type aliases?

Reply:
I haven't thought of that. I don't think it will just work, but I think it can be made to work.

Associated types will be type parameters of the trait and the trait's methods, like this:
```rust
fn foo<Self : MyTrait<T>, T>(self) -> T
```
So we have to ensure that whatever makes `Self` a `MyTrait` has a type argument for `T` and we should read that off of the default if nothing else is provided. So for an `impl` block that uses the default it should behave as if the default was written for `T`:
```rust
impl MyTrait<Option<Foo>> for Foo { ... }
```